torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -597,6 +774,232 @@ static void PrepareForRemoveMutations(MutationRemover& mr, Block* b) {
+      origValue, cloneNode->output(), block, next_node, output_name);
+
+  return MatchIfBlocksOutputForValue(origValue, block, cloneNode->output());
+}
+
+Value* findArgumentAsInputParam(
+    const std::shared_ptr<Graph>& graph,
+    std::string& name,
+    IValue& attr) {
+  for (auto input : graph->inputs()) {
+    if ((input->debugName() == name))
+      return input;
+  }
+  throw std::runtime_error(
+      "Attribute is not part of model parameters. Cannot handle SetAttr and GetAttr nodes for : " +
+      name);
+}
+
+std::unordered_map<std::string, Value*> registerInplaceOpAsBlockOutputs(
+    MutationRemover& mr,
+    Block* block,
+    const std::shared_ptr<Graph>& graph,
+    Module module_,
+    std::unordered_map<std::string, Value*> setAttrValues) {
+  Node* m = *block->nodes().begin();
+  WithInsertPoint guard(m);
+
+  std::unordered_map<std::string, Value*> nextSetAttrValues = {};
+
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    Node* n = *it;
+    it++; // node n can be destroyed
+
+    if (n->kind() == prim::GetAttr || n->kind() == prim::SetAttr) {
+      auto name = n->s(attr::name);
+      auto attrModule = module_;
+
+      auto moduleNames =
+          findSubModuleAttr(n->inputs().at(0), name, attrModule, graph);
+      if (!attrModule.hasattr(name))
+        continue;
+      auto attr = attrModule.attr(name);
+      Value* paramConst = nullptr;
+
+      std::string fullName("");
+      for (auto& name : moduleNames) {
+        fullName += name + '.';
+      }
+      fullName += name;
+
+      auto type = attrModule.type();
+      auto slot = *type->findAttributeSlot(name);
+
+      // Add model_parameters and model_buffers as model inputs. Order is
+      // preserved based on the appearance in the graph.
+      if (type->is_parameter(slot) || type->is_buffer(slot)) {
+        if (attrValues.find(fullName) != attrValues.end()) {
+          paramConst = attrValues[fullName];
+        } else {
+          paramConst = findArgumentAsInputParam(graph, fullName, attr);
+          if (!paramConst) {
+            toDestory.emplace_back(n);
+            continue;
+          }
+          attrValues.insert({fullName, paramConst});
+        }
+      }
+
+      if (n->kind() == prim::SetAttr) { // Handle SetAttr node
+        if (attrModule.hasattr(name)) {
+          // If inside a block, keep the output value to register in block
+          // output.
+          auto block_ = n->owningBlock();
+          if (block_->owningNode() &&
+              block_->owningNode()->kind() == prim::If) { // TODO: Add loop
+
+            auto attrValue =
+                (setAttrValues.find(fullName) != setAttrValues.end())
+                ? setAttrValues[fullName]
+                : attrValues[fullName];
+
+            auto blockOutput = registerSetAttrInBlocks(
+                graph, block_, n->inputs().at(1), attrValue, fullName);
+            nextSetAttrValues[fullName] = blockOutput;
```

Comment:
why differentiate `nextSetAttrValues` and `setAttrValues`?

Reply:
The point is that when you are in the block, the output of setAttrNode should be considered as the latest setAttrValue.
But as soon as you return from the block, the output of the block should be considered as the latest setAttrValue.
And, we should record the latest setAttrValue as we enter a new block in the depth-first traversal.